org.jacoco.core/src/org/jacoco/core/data/ExecutionData.java

```diff
@@ -159,10 +158,25 @@ public void merge(final ExecutionData other) {
 	public void merge(final ExecutionData other, final boolean flag) {
 		assertCompatibility(other.getId(), other.getName(),
 				other.getProbes().length);
-		final boolean[] otherData = other.getProbes();
+		final int[] otherData = other.getProbes();
 		for (int i = 0; i < probes.length; i++) {
-			if (otherData[i]) {
-				probes[i] = flag;
+			int otherProbe = otherData[i];
+			if (otherProbe > 0) {
+				if (flag) {
+					int sum = probes[i] + otherProbe;
+					if (sum == Integer.MAX_VALUE || sum < 0) {
```

Comment:
Hmm I see. So that alternative achieves an upper bound through casting to and from a larger data type. I'm not sure how much this `merge` method impacts performance as I suspect most of the performance loss is in the injected bytecode (see `ProbeInserter`). Anyway, I'll try to clarify my thoughts on the implementation here (and once we agree on the implementation I'll update the comment to clarify it as well). 

If I understand correctly, the `ExecutionData` holds the `int[]` reference that is also injected in the instrumented class. My upper bound for values in this array is `Integer.MAX_VALUE - 1` (see also the injected bytecode in `ProbeInserter`) because I don't want the integer to overflow once we increment. If I would allow `Integer.MAX_VALUE` and increment it, then `Math.min(max-value-incremented-integer, MAX_VALUE -1)` (in the injected bytecode) will return the integer because it is the value `Integer.MIN_VALUE` since it overflowed. I can't use a `Math.max` implementation either because the probes start with value 0. For that reason, I also don't allow the values in the probe array to become `MAX_VALUE` here as otherwise they would still overflow. I hope this clarifies it a bit. 

We can, however, look at changing the injected bytecode to do this branch-free alternative. If we are able to do that, we can also do this branch-free alternative elsewhere in the code too. I will need to look at implementing it in bytecode myself without the library as the library may not be on the classpath, but that shouldn't be so hard.

Reply:
As discussed, I have been able to get an implementation working with double casting, such that the array is incremented as follows:

```
ALOAD <array>
ICONST <index>
DUP2
IALOAD
I2D
DCONST_1
DADD
D2I
IASTORE
```

I have rerun my performance tests (20 runs) with the base now on `0.8.7` and the applied performance improvements in other comments. The `Math.min` implementation has shown a performance increase of ~3.5% and the `double-cast` implementation has shown a performance increase of ~2.5%. 

Shall I apply the double casting implementation on this PR or do you wish to start a whole new PR on a new base branch based on 0.8.7 and do another run?